README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# bootwar

<!-- badges: start -->
<!-- badges: end -->

‘bootwar’ aims to help users build their intuition about nonparametric
bootstrap tests with pooled resampling, especially for independent and
paired t-tests (Dwivedi, Mallawaarachchi, and Alvarado, 2017).

Inspired by the card game War, which has straightforward rules but can
be time-consuming, ‘bootwar’ is a variant with smaller sample sizes.
Here, after playing a user-defined number of rounds, a winner is
determined through nonparametric bootstrap tests with pooled resampling
as implemented in the
[npboottprm](https://CRAN.R-project.org/package=npboottprm) R package.

## Installation

You can install the released version of ‘bootwar’ from
[CRAN](https://CRAN.R-project.org):

``` r
install.packages("bootwar")
```

Install the development version of bootwar from GitHub using
[devtools](https://devtools.r-lib.org/):

``` r
# install.packages("devtools")
devtools::install_github("mightymetrika/bootwar")
```

## Example

This example demonstrates how to play a five-round game of bootwar:

### Initialization:

``` r
# Load bootwar
library(bootwar)

# Set up vectors for computer and player's cards and values
comp_cv <- vector(mode = "character")
comp_vv <- vector(mode = "numeric")
plyr_cv <- vector(mode = "character")
plyr_vv <- vector(mode = "numeric")

# Shuffle a 52 card standard deck from the 'mmcards' package
sdeck <- mmcards::shuffle_deck(seed = 150)
head(sdeck)
#>    rank suit card value
#> 35   10    H  10H 10.50
#> 25    K    D   KD 13.25
#> 31    6    H   6H  6.50
#> 5     6    C   6C  6.00
#> 27    2    H   2H  2.50
#> 44    6    S   6S  6.75
```

### Play the First Round:

``` r
rres <- play_round(cdeck = sdeck,
                   plyr_cv = plyr_cv, plyr_vv = plyr_vv,
                   comp_cv = comp_cv, comp_vv = comp_vv)
```

### Continue the Game for Four More Rounds:

``` r
for (i in 1:4) {
  rres <- play_round(cdeck = rres$updated_deck,
                     plyr_cv = rres$plyr_cv, plyr_vv = rres$plyr_vv,
                     comp_cv = rres$comp_cv, comp_vv = rres$comp_vv)
}

# Ensure 10 cards have been dealt
nrow(rres$updated_deck)
#> [1] 42
```

### Analyze the Game:

Use analyze_game() to evaluate the results. In this example, we’ll apply
the paired t-test, but users can opt for the independent t-test by
setting mode to ‘t’. Modify the conf.level parameter to change the
confidence interval (think of this as a level which controls the amount
of draws).

``` r
gres <- analyze_game(plyr_vv = rres$plyr_vv, comp_vv = rres$comp_vv,
                     mode = "pt", nboot = 1000, seed = 150, conf.level = 0.05)

# Display game results
gres$winner
#> [1] "Computer Wins"
gres$bootstrap_results$effect.size
#> [1] -0.25
gres$bootstrap_results$ci.effect.size
#> 47.5% 52.5% 
#> -0.15  0.05
gres$bootstrap_results$p.value
#> [1] 0.838
```

For a more streamlined gameplay experience, the bootwar() function
launches an interactive Shiny web application.

Users can also customize the game using an anonymous function for deck
definition, as showcased in the anonymous-bootwar vignette. Users can
also customize the game by using an anonymous function to define a
different deck for each player, as showcased in the interleaved-bootwar
vignette.

## References

Dwivedi AK, Mallawaarachchi I, Alvarado LA (2017). “Analysis of small
sample size studies using nonparametric bootstrap test with pooled
resampling method.” Statistics in Medicine, 36 (14), 2187-2205.


NEWS.md
# bootwar (development version)

# bootwar 0.2.1

* Updated DESCRIPTION

# bootwar 0.2.0

* Incorprated 'mmcards'to handle playing card functionality.

* Added pictures to Standard Deck game play.


# bootwar 0.1.0

* Added a `NEWS.md` file to track changes to the package.



tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(bootwar)

test_check("bootwar")


tests/testthat/test-game_mgmt.R
test_that("score keeper works as expected", {

  # Set up vectors
  comp_cv <- vector(mode = "character")
  comp_vv <- vector(mode = "numeric")
  plyr_cv <- vector(mode = "character")
  plyr_vv <- vector(mode = "numeric")


  # Return a shuffled_deck
  shd <- mmcards::shuffle_deck(seed = 150)

  #Round 1

  # Deal first card to comp
  card <- mmcards::deal_card(shd)
  comp_cv <- c(comp_cv, card$dealt_card$card)
  comp_vv <- c(comp_vv, card$dealt_card$value)

  # Deal first card to plyr
  card <- mmcards::deal_card(card$updated_deck)
  plyr_cv <- c(plyr_cv, card$dealt_card$card)
  plyr_vv <- c(plyr_vv, card$dealt_card$value)

  # Get round 1 scores
  scores <- score_keeper(plyr_vv, comp_vv, mode = "t")

  expect_equal(scores$effect_size, 2.75)

  # Round 2

  card <- mmcards::deal_card(card$updated_deck)
  comp_cv <- c(comp_cv, card$dealt_card$card)
  comp_vv <- c(comp_vv, card$dealt_card$value)

  card <- mmcards::deal_card(card$updated_deck)
  plyr_cv <- c(plyr_cv, card$dealt_card$card)
  plyr_vv <- c(plyr_vv, card$dealt_card$value)

  # Get round 2 scores
  scores <- score_keeper(plyr_vv, comp_vv, mode = "t")

  expect_equal(scores$effect_size, 1.125)

})

test_that("play round works as expected", {

  # Set up vectors
  comp_cv <- vector(mode = "character")
  comp_vv <- vector(mode = "numeric")
  plyr_cv <- vector(mode = "character")
  plyr_vv <- vector(mode = "numeric")


  # Return a shuffled_deck
  ideck <- mmcards::shuffle_deck(seed = 150)

  # Round 1
  rres <- play_round(cdeck = ideck, plyr_cv = plyr_cv, plyr_vv = plyr_vv,
                     comp_cv = comp_cv, comp_vv = comp_vv)

  # Get Round 1 scores
  scores <- score_keeper(rres$plyr_vv, rres$comp_vv, mode = "t")
  expect_equal(scores$effect_size, -2.75)

  # Round 2
  rres <- play_round(cdeck = rres$updated_deck, plyr_cv = rres$ply_cv, plyr_vv = rres$plyr_vv,
                     comp_cv = rres$comp_cv, comp_vv = rres$comp_vv)

  # Get round 2 scores
  scores <- score_keeper(rres$plyr_vv, rres$comp_vv, mode = "t")

  expect_equal(scores$effect_size, -1.125)

})


tests/testthat/test-deck_funcs.R
test_that("shuffled deck works with deck", {

  # Decks equal with same seed
  shd1 <- shuffle_deck(seed = 100)
  shd2 <- shuffle_deck(seed = 100)


  expect_equal(shd1$value[[1]], shd2$value[[1]])
  expect_equal(shd1$value[[13]], shd2$value[[13]])
  expect_equal(shd1$value[[26]], shd2$value[[26]])
  expect_equal(shd1$value[[39]], shd2$value[[39]])
  expect_equal(shd1$value[[52]], shd2$value[[52]])
})

test_that("shuffled deck works with anonymous deck", {

  # Decks equal with same seed
  shd1 <- shuffle_deck(deck_of_cards = function(x)stats::rnorm(52, mean = 8.375, sd = 3.788689),
                       seed = 100)

  expect_equal(round(mean(shd1$value), 2), 8.51)
})

test_that("shuffled deck works with interleaved deck", {

  # Decks equal with same seed
  shd1 <- shuffle_deck(deck_of_cards = function(x) {list(stats::rnorm(26, 1, 2),
                                                         stats::rnorm(26, 1.5, 1.5))},
                       seed = 100)

  expect_equal(round(mean(shd1$value), 2), 1.33)
})

test_that("shuffled deck throws error with interleaved deck with unequal size", {
  expect_error(shuffle_deck(deck_of_cards = function(x) {list(stats::rnorm(26, 1, 2),
                                                              stats::rnorm(25, 1.5, 1.5))},
                            seed = 100))
})

test_that("deal works with a shuffled deck", {

  # Deal two cards from the same deck
  shd1 <- shuffle_deck(seed = 100)
  rshd1 <- nrow(shd1)
  card <- deal_card(shd1)
  shd1 <- card$updated_deck
  rshd2 <- nrow(shd1)
  card2 <- deal_card(shd1)

  # Test that we are holding card values
  expect_equal(card$dealt_card$value, 4.25)
  expect_equal(card2$dealt_card$value, 11.25)

  # Test that our deck is shrinking
  expect_equal(rshd1, rshd2 + 1)
})


R/game_mgmt.R
#' Calculate Scores and Effect Size
#'
#' This function computes the sum and mean of the player's and computer's values
#' and calculates the effect size based on the given mode (`t` or `pt`).
#'
#' @param player_values A numeric vector representing the values of the player's
#'    cards.
#' @param comp_values A numeric vector representing the values of the computer's
#'    cards.
#' @param mode A character string representing the mode of the game, either 't'
#'    for independent t-test or 'pt' for paired t-test.
#'
#' @return A list containing:
#' \itemize{
#'   \item \code{player_sum}: Sum of player's values.
#'   \item \code{player_mean}: Mean of player's values.
#'   \item \code{comp_sum}: Sum of computer's values.
#'   \item \code{comp_mean}: Mean of computer's values.
#'   \item \code{effect_size}: Calculated effect size based on the given mode.
#' }
#'
#' @examples
#' # Calculate scores for a simple game
#' player_vals <- c(2.5, 3.0, 4.5)
#' comp_vals <- c(3.5, 2.0, 4.0)
#' scores <- score_keeper(player_vals, comp_vals, mode = "t")
#'
#' @export
score_keeper <- function(player_values, comp_values, mode) {
  # Check for valid mode
  if (!mode %in% c('t', 'pt')) {
    stop("Invalid mode. Please use 't' or 'pt'.")
  }

  # Calculate basic stats
  player_sum <- sum(player_values)
  player_mean <- mean(player_values)

  comp_sum <- sum(comp_values)
  comp_mean <- mean(comp_values)

  # Calculate effect size
  effect_size <- NULL
  if (mode == 't') {
    effect_size <- player_mean - comp_mean
  } else if (mode == 'pt') {
    if (length(player_values) != length(comp_values)) {
      stop("For paired t-test, player and comp values should have the same length.")
    }
    effect_size <- mean(player_values - comp_values)
  }

  # Return results as a list
  result <- list(
    player_sum = player_sum,
    player_mean = player_mean,
    comp_sum = comp_sum,
    comp_mean = comp_mean,
    effect_size = effect_size
  )

  return(result)
}

#' Play a Round of the Card Game
#'
#' This function simulates a single round of the card game, where both the
#' computer and the player are dealt a card. The function returns the updated
#' state of the game after the round.
#'
#' @param cdeck A dataframe representing the current deck of cards.
#' @param plyr_cv A character vector storing the cards dealt to the player so far.
#' @param plyr_vv A numeric vector storing the values of the cards dealt to the
#'    player so far.
#' @param plyr_ic A character vector storing the image cards dealt to the player. Default is NULL.
#' @param comp_cv A character vector storing the cards dealt to the computer so
#'    far.
#' @param comp_vv A numeric vector storing the values of the cards dealt to the
#'    computer so far.
#' @param comp_ic A character vector storing the image cards dealt to the computer. Default is NULL.
#'
#' @return A list containing:
#' \itemize{
#'   \item \code{updated_deck}: A dataframe representing the updated deck of cards after the round.
#'   \item \code{plyr_cv}: Updated character vector of cards dealt to the player.
#'   \item \code{plyr_vv}: Updated numeric vector of values of cards dealt to the player.
#'   \item \code{plyr_ic}: Updated character vector of image cards dealt to the player.
#'   \item \code{comp_cv}: Updated character vector of cards dealt to the computer.
#'   \item \code{comp_vv}: Updated numeric vector of values of cards dealt to the computer.
#'   \item \code{comp_ic}: Updated character vector of image cards dealt to the computer.
#' }
#'
#' @examples
#' # Simulate a round of the game with a sample deck
#' deck <- mmcards::shuffle_deck()
#' plyr_cards <- character(0)
#' plyr_values <- numeric(0)
#' comp_cards <- character(0)
#' comp_values <- numeric(0)
#' round_result <- play_round(deck, plyr_cv = plyr_cards, plyr_vv = plyr_values,
#'                            comp_cv = comp_cards, comp_vv = comp_values)
#'
#' @export
play_round <- function(cdeck, plyr_cv, plyr_vv, plyr_ic = NULL, comp_cv, comp_vv, comp_ic = NULL) {

  # Deal card to player
  card_plyr <- mmcards::deal_card(cdeck)
  plyr_cv <- c(plyr_cv, as.character(card_plyr$dealt_card$card))
  plyr_vv <- c(plyr_vv, card_plyr$dealt_card$value)
  if(inherits(card_plyr$updated_deck, "ImgDeck")){
    plyr_ic <- c(plyr_ic, card_plyr$dealt_card$icard)
  } else {
    plyr_ic <- c(plyr_ic, NULL)
  }


  # Update deck after comp's card is dealt
  shd <- card_plyr$updated_deck

  # Deal card to comp
  card_comp <- mmcards::deal_card(shd)
  comp_cv <- c(comp_cv, as.character(card_comp$dealt_card$card))
  comp_vv <- c(comp_vv, card_comp$dealt_card$value)
  if (inherits(card_comp$updated_deck, "ImgDeck")){
    comp_ic <- c(comp_ic, card_comp$dealt_card$icard)
  } else {
    comp_ic <- c(comp_ic, NULL)
  }


  # Return updated state
  return(list(updated_deck = card_comp$updated_deck,
              plyr_cv = plyr_cv, plyr_vv = plyr_vv, plyr_ic = plyr_ic,
              comp_cv = comp_cv, comp_vv = comp_vv, comp_ic = comp_ic))
}



R/deck_funcs.R
#' Shuffle a Deck of Cards
#'
#' This function shuffles a given deck of cards. If the provided deck is a function,
#' it generates values based on the anonymous function before shuffling.
#' If not, it directly shuffles the provided deck.
#'
#' @param deck_of_cards A dataframe representing the deck of cards or a function
#'    that generates values for the deck. Defaults to a predefined deck named `deck`.
#' @param seed An optional seed for reproducibility.
#'
#' @return A shuffled dataframe of cards. The dataframe will have a class attribute
#'         of "data.frame", "shuffled deck", and either "deck", "anonymous deck"
#'         or "interleaved deck" depending on the input.
#'
#' @examples
#' # Using a predefined deck
#' shuffled <- shuffle_deck(seed = 123)
#'
#' # Using an anonymous function
#' fun_deck <- function(x) { stats::runif(52, 1, 52) }
#' shuffled <- shuffle_deck(deck_of_cards = fun_deck, seed = 123)
#'
#' # Using an anonymous function to create an interleaved deck
#' shuffled <- shuffle_deck(deck_of_cards = function(x) {list(stats::rnorm(26, 1, 2),
#'                                                            stats::rnorm(26, 1.5, 1.5))},
#'                          seed = 100)
#'
#' @export
shuffle_deck <- function(deck_of_cards = deck, seed) {

  # Switch out globals
  if (is.function(deck_of_cards)) deck <- NULL

  # Set seed for reproducibility
  if (!is.null(seed)) {
    set.seed(seed)
  }

  # Check if deck_of_cards is a function
  if (is.function(deck_of_cards)) {
    # Generate the values based on the anonymous function
    decks <- deck_of_cards(NULL)

    # If the result is a list with two elements, treat them as separate decks
    if (is.list(decks) && length(decks) == 2) {

      # Check if the two decks are of the same length
      if (length(decks[[1]]) != length(decks[[2]])) {
        stop("Both decks must be of the same length for interleaving!")
      }

      # Create separate decks for player and computer
      A_deck <- data.frame(card = paste("A", 1:length(decks[[1]]), sep = "_"), value = decks[[1]])
      B_deck <- data.frame(card = paste("B", 1:length(decks[[2]]), sep = "_"), value = decks[[2]])

      # Shuffle each deck
      A_deck <- A_deck[sample(nrow(A_deck)), ]
      B_deck <- B_deck[sample(nrow(B_deck)), ]

      # Interleave the decks
      interleaved_deck <- rbind(A_deck, B_deck)[order(rep(1:nrow(A_deck), 2)), ]

      class(interleaved_deck) <- c("data.frame", "shuffled deck", "interleaved deck")
      return(interleaved_deck)

    } else {
      # If it's just one deck, treat it as before
      values <- decks
      ordered_indices <- order(values, decreasing = FALSE)
      ordered_deck <- data.frame(card = 1:length(values), value = values[ordered_indices])

      # Shuffle the ordered deck
      shuffled_deck <- ordered_deck[sample(nrow(ordered_deck)), ]
      class(shuffled_deck) <- c("data.frame", "shuffled deck", "anonymous deck")
      return(shuffled_deck)
    }
  } else {
    # Shuffle the provided deck
    shuffled_deck <- deck_of_cards[sample(nrow(deck_of_cards)), ]
    class(shuffled_deck) <- c("data.frame", "shuffled deck", "deck")
    return(shuffled_deck)
  }
}


#' Deal a Card from the Deck
#'
#' This function deals the top card from a given deck and updates the deck by
#' removing the dealt card. If the deck is empty, the function will return an
#' error indicating there are no more cards to deal.
#'
#' @param current_deck A dataframe representing the current deck of cards.
#'
#' @return A list containing two elements:
#' \itemize{
#'   \item \code{dealt_card}: A dataframe containing the dealt card.
#'   \item \code{updated_deck}: A dataframe representing the updated deck after dealing.
#' }
#'
#' @examples
#' # Create a sample deck
#' deck <- data.frame(card = c("A", "B", "C"), value = c(1, 2, 3))
#'
#' # Deal a card from the sample deck
#' result <- deal_card(deck)
#' dealt <- result$dealt_card
#' remaining <- result$updated_deck
#'
#' @export
deal_card <- function(current_deck) {
  if (nrow(current_deck) == 0) {
    stop("No more cards in the deck!")
  }

  card_to_deal <- current_deck[1, ]
  updated_deck <- current_deck[-1, ]  # Remove the top card

  return(list(dealt_card = card_to_deal, updated_deck = updated_deck))
}


R/app.R
#' Bootwar Shiny App
#'
#' Launches a Shiny application for the Bootwar card game.
#' The app allows users to play a card game where they can analyze the game
#' results using nonparametric bootstrap test with pooled resampling methods.
#'
#' @details
#' The Bootwar card game is a bootstrap variation of the card game War. The
#' Bootwar application has options to select different modes ('t' for
#' independent t-test and 'pt' for paired t-test) and decks.
#' Players can use a standard 52 card deck and they can also input a custom
#' anonymous function to generate a deck. The app will let users deal cards,
#' play the game, and then score and analyze results using nonparametric
#' bootstrap test with pooled resampling methods. The game is designed to help
#' users gain greater intuition on nonparametric bootstrap test with pooled
#' resampling methods; as such, players are encouraged to experiment with
#' different confidence levels, number of rounds, number of bootstrap resamples,
#' and custom decks.
#'
#' @return
#' A Shiny application object. Running this function will launch the Shiny app
#' in the user's default web browser.
#'
#' @examples
#' if(interactive()){
#'   bootwar()
#' }
#'
#' @export
bootwar <- function(){

  ui <- shiny::fluidPage(
    theme = shinythemes::shinytheme("united"),
    shinyjs::useShinyjs(),  # Initialize shinyjs
    shiny::titlePanel("Bootwar Game"),

    # Settings Panel
    shiny::sidebarLayout(
      shiny::sidebarPanel(
        shiny::selectInput("mode", "Select Mode:", choices = c("t", "pt")),
        shiny::selectInput("deck", "Select Deck:", choices = c("Standard", "Anonymous")),
        shiny::conditionalPanel(
          condition = "input.deck === 'Anonymous'",
          shiny::textInput("anon_func", "Enter Anonymous Function:", value = "function(x) { runif(52, 1, 52) }"),
          shiny::actionButton("eval_anon_func", "Evaluate Anonymous Function")
        ),
        shiny::numericInput("conf.level", "Confidence Level:", 0.95),
        shiny::numericInput("nboot", "Number of Bootstrap Resamples:", 1000),
        shiny::numericInput("seed", "Seed:", NA),
        shiny::sliderInput("rounds", "Number of Rounds:", min = 1, max = 26, value = 5),
        shiny::actionButton("new_game", "Start New Game"),
        shinyjs::disabled( shiny::actionButton("deal_card", "Deal Card")),
        shiny::verbatimTextOutput("round_counter"), # New round counter output
      ),

      # Gameplay and Results Area
      shiny::mainPanel(
        shiny::fluidRow(
          # Player Column
          shiny::column(6,
                        shiny::h3("Player"),
                        shiny::conditionalPanel(
                          condition = "input.deck === 'Anonymous'",
                          shiny::verbatimTextOutput("player_card")
                        ),
                        shiny::conditionalPanel(
                          condition = "input.deck === 'Standard'",
                          shiny::imageOutput("player_card_image", width = "100px", height="auto")
                        ),
                        shiny::br(),  # Add a line break
                        shiny::verbatimTextOutput("player_value"),
                        shiny::verbatimTextOutput("player_running_sum"),
                        shiny::verbatimTextOutput("player_running_mean")
          ),

          # Computer Column
          shiny::column(6,
                        shiny::h3("Computer"),
                        shiny::conditionalPanel(
                          condition = "input.deck === 'Anonymous'",
                          shiny::verbatimTextOutput("comp_card")
                        ),
                        shiny::conditionalPanel(
                          condition = "input.deck === 'Standard'",
                          shiny::imageOutput("comp_card_image", width = "100px", height="auto")
                        ),
                        shiny::br(),  # Add a line break
                        shiny::verbatimTextOutput("comp_value"),
                        shiny::verbatimTextOutput("comp_running_sum"),
                        shiny::verbatimTextOutput("comp_running_mean")
          )
        ),

        shiny::h3("Score Board"),
        shiny::verbatimTextOutput("effect_stats"),
        shiny::uiOutput("resultsUI")
      )
    )
  )

  server <- function(input, output, session) {
    # Switch out globals
    x <- NULL

    # Game state
    game_state <- shiny::reactiveVal(list(deck = NULL, player_values = numeric(0), comp_values = numeric(0), current_round = 0))

    shiny::observeEvent(input$eval_anon_func, {

      # Try to evaluate the entered R code to get the anonymous function
      tryCatch({
        user_func <- eval(parse(text = input$anon_func))
        if (is.function(user_func)) {
          # Generate the custom deck
          custom_deck <- mmcards::shuffle_deck(deck_of_cards = user_func, seed = process_seed(input$seed))

          # Update game state with this custom deck
          current_state <- game_state()
          current_state$deck <- custom_deck
          game_state(current_state)

          shiny::showNotification("Anonymous function evaluated successfully and deck created!", type = "message")
        } else {
          shiny::showNotification("Provided input is not a valid anonymous function.", type = "error")
        }
      }, error = function(e) {
        shiny::showNotification(paste("Error:", e$message), type = "error")
      })
    })


    shiny::observeEvent(input$new_game, {

      # Shuffle the deck
      if (input$deck == "Anonymous" && !is.null(input$anon_func)) {
        # If Anonymous is selected and there is a previously evaluated function,
        # shuffle the deck using the anonymous function
        user_func <- tryCatch(eval(parse(text = input$anon_func)), error = function(e) NULL)
        if (is.function(user_func)) {
          current_deck <- mmcards::shuffle_deck(deck_of_cards = user_func, seed = process_seed(input$seed))
        } else {
          # Default to a standard deck if the function evaluation fails
          current_deck <- mmcards::i_deck(deck = mmcards::shuffle_deck(seed = process_seed(input$seed)),
                                          i_path = "www",
                                          i_names = c("2_of_clubs", "2_of_diamonds", "2_of_hearts", "2_of_spades",
                                                      "3_of_clubs", "3_of_diamonds", "3_of_hearts", "3_of_spades",
                                                      "4_of_clubs", "4_of_diamonds", "4_of_hearts", "4_of_spades",
                                                      "5_of_clubs", "5_of_diamonds", "5_of_hearts", "5_of_spades",
                                                      "6_of_clubs", "6_of_diamonds", "6_of_hearts", "6_of_spades",
                                                      "7_of_clubs", "7_of_diamonds", "7_of_hearts", "7_of_spades",
                                                      "8_of_clubs", "8_of_diamonds", "8_of_hearts", "8_of_spades",
                                                      "9_of_clubs", "9_of_diamonds", "9_of_hearts", "9_of_spades",
                                                      "10_of_clubs", "10_of_diamonds", "10_of_hearts", "10_of_spades",
                                                      "jack_of_clubs", "jack_of_diamonds", "jack_of_hearts", "jack_of_spades",
                                                      "queen_of_clubs", "queen_of_diamonds", "queen_of_hearts", "queen_of_spades",
                                                      "king_of_clubs", "king_of_diamonds", "king_of_hearts", "king_of_spades",
                                                      "ace_of_clubs", "ace_of_diamonds", "ace_of_hearts", "ace_of_spades2"
                                          ))

        }
      } else {
        # Shuffle the standard deck if Standard is selected
        current_deck <- mmcards::i_deck(deck = mmcards::shuffle_deck(seed = process_seed(input$seed)),
                                        i_path = "www",
                                        i_names = c("2_of_clubs", "2_of_diamonds", "2_of_hearts", "2_of_spades",
                                                    "3_of_clubs", "3_of_diamonds", "3_of_hearts", "3_of_spades",
                                                    "4_of_clubs", "4_of_diamonds", "4_of_hearts", "4_of_spades",
                                                    "5_of_clubs", "5_of_diamonds", "5_of_hearts", "5_of_spades",
                                                    "6_of_clubs", "6_of_diamonds", "6_of_hearts", "6_of_spades",
                                                    "7_of_clubs", "7_of_diamonds", "7_of_hearts", "7_of_spades",
                                                    "8_of_clubs", "8_of_diamonds", "8_of_hearts", "8_of_spades",
                                                    "9_of_clubs", "9_of_diamonds", "9_of_hearts", "9_of_spades",
                                                    "10_of_clubs", "10_of_diamonds", "10_of_hearts", "10_of_spades",
                                                    "jack_of_clubs", "jack_of_diamonds", "jack_of_hearts", "jack_of_spades",
                                                    "queen_of_clubs", "queen_of_diamonds", "queen_of_hearts", "queen_of_spades",
                                                    "king_of_clubs", "king_of_diamonds", "king_of_hearts", "king_of_spades",
                                                    "ace_of_clubs", "ace_of_diamonds", "ace_of_hearts", "ace_of_spades2"
                                        ))
      }

      game_state(list(deck = current_deck, player_values = numeric(0), comp_values = numeric(0), current_round = 0))

      # Enable the "Deal Card" button when a new game starts
      shinyjs::enable("deal_card")
    })

    shiny::observeEvent(input$deal_card, {
      # Get current state
      current_state <- game_state()

      # Play the round
      round_result <- play_round(cdeck = current_state$deck,
                                 plyr_cv = current_state$player_cards, plyr_vv = current_state$player_values, plyr_ic = current_state$player_icards,
                                 comp_cv = current_state$comp_cards, comp_vv = current_state$comp_values, comp_ic = current_state$comp_icards)

      # Calculate scores
      scores <- score_keeper(round_result$plyr_vv, round_result$comp_vv, mode = input$mode)

      # Update game state
      game_state(list(deck = round_result$updated_deck,
                      player_cards = round_result$plyr_cv,
                      player_values = round_result$plyr_vv,
                      player_icards = round_result$plyr_ic,
                      comp_cards = round_result$comp_cv,
                      comp_values = round_result$comp_vv,
                      comp_icards = round_result$comp_ic,
                      player_scores = scores,
                      current_round = current_state$current_round + 1))

      # Check if it's the last round
      if (game_state()$current_round == input$rounds) {
        # Analyze game
        analysis <- game_analysis()  # Use the eventReactive result

        # Update game state with the analysis
        current_game_state <- game_state()
        current_game_state$analysis <- analysis
        game_state(current_game_state)

        # Notify user
        shiny::showNotification("Game has ended! Check out the results below.", type = "message", duration = NULL)

        # Disable the "Deal Card" button when the game ends
        shinyjs::disable("deal_card")
      }
    })

    # 1. Use eventReactive to determine when game is complete
    game_analysis <- shiny::eventReactive(game_state()$current_round == input$rounds, {

      analyze_game(game_state()$player_values, game_state()$comp_values,
                   mode = input$mode, conf.level = input$conf.level,
                   nboot = input$nboot, seed = process_seed(input$seed))
    })

    # Update UI based on game state

    # Update the UI with the round counter
    output$round_counter <- shiny::renderText({
      paste("Current Round:", game_state()$current_round, "of", input$rounds)
    })
    output$player_card <- shiny::renderText({
      shiny::req(input$deck == "Anonymous")
      paste0("Card: ", { utils::tail(game_state()$player_cards, 1) })
    })
    output$comp_card <- shiny::renderText({
      shiny::req(input$deck == "Anonymous")
      paste0("Card: ", { utils::tail(game_state()$comp_cards, 1) })
    })
    output$player_value <- shiny::renderText(paste0("Card Value: ", { utils::tail(game_state()$player_values, 1) }))
    output$comp_value <- shiny::renderText(paste0("Card Value: ", { utils::tail(game_state()$comp_values, 1) }))

    # Render player card image
    output$player_card_image <- shiny::renderImage({
      # Ensure that there are cards in the player_icards list
      shiny::req(length(game_state()$player_icards) > 0)

      icard <- utils::tail(game_state()$player_icards, 1)

      list(src = system.file(icard, package = "bootwar"), contentType = "image/png", width=200, height="auto", alt = utils::tail(game_state()$player_cards, 1))
    }, deleteFile = FALSE)

    # Render computer card image
    output$comp_card_image <- shiny::renderImage({
      # Ensure that there are cards in the comp_icards list
      shiny::req(length(game_state()$comp_icards) > 0)

      icard <- utils::tail(game_state()$comp_icards, 1)

      list(src = system.file(icard, package = "bootwar"), contentType = "image/png", width=200, height="auto", alt = utils::tail(game_state()$comp_cards, 1))
    }, deleteFile = FALSE)

    output$player_running_sum <- shiny::renderText({
      scores <- game_state()$player_scores
      paste0("Running Sum: ", scores$player_sum)
    })

    output$player_running_mean <- shiny::renderText({
      scores <- game_state()$player_scores
      formatted_mean <- formatC(scores$player_mean, format = "f", digits = 2)
      paste0("Running Mean: ", formatted_mean)
    })

    output$comp_running_sum <- shiny::renderText({
      scores <- game_state()$player_scores
      paste0("Running Sum: ", scores$comp_sum)
    })

    output$comp_running_mean <- shiny::renderText({
      scores <- game_state()$player_scores
      formatted_mean <- formatC(scores$comp_mean, format = "f", digits = 2)
      paste0("Running Mean: ", formatted_mean)
    })

    output$effect_stats <- shiny::renderText({
      scores <- game_state()$player_scores
      formatted_effect_size <- formatC(scores$effect_size, format = "f", digits = 2)
      paste0("Running Effect Size: ", formatted_effect_size)
    })

    # Update UI with results from analyze_game (assuming it's stored in the game_state)
    output$winner <- shiny::renderText({
      game_state()$analysis$winner
    })

    output$test_stats <- shiny::renderTable({
      shiny::req(game_state()$analysis$bootstrap_results, game_state()$analysis$classical_results)

      # Bootstrap results
      boot_res <- game_state()$analysis$bootstrap_results

      # Classical test results
      classical_res <- game_state()$analysis$classical_results

      df_res_stat <- data.frame(
        Test_Statistic = boot_res$orig.stat,
        CI_Lower = boot_res$ci.stat[1],
        CI_Upper = boot_res$ci.stat[2],
        Bootstrap_P_Value = boot_res$p.value,
        Classical_P_Value = classical_res$p.value
      )

      names(df_res_stat) <- c("Test Statistic", "Bootstrap CI (Lower)", "Bootstrap CI (Upper)",
                              "Bootstrap P-value", "Welch P-value")
      df_res_stat
    }, row.names = FALSE)

    output$effect_size <- shiny::renderTable({
      shiny::req(game_state()$analysis$bootstrap_results)

      # Bootstrap results
      boot_res <- game_state()$analysis$bootstrap_results

      df_res_effect <- data.frame(
        Effect_Size = boot_res$effect.size,
        CI_Lower = boot_res$ci.effect.size[1],
        CI_Upper = boot_res$ci.effect.size[2]
      )

      names(df_res_effect) <- c("Effect Size", "CI (Lower)", "CI (Upper)")
      df_res_effect
    }, row.names = FALSE)

    # Test Statistic Histogram
    output$hist_stat <- shiny::renderPlot({
      shiny::req(game_state()$analysis$bootstrap_results)

      # Bootstrap results
      boot_res <- game_state()$analysis$bootstrap_results

      ggplot2::ggplot(data.frame(x = boot_res$bootstrap.stat.dist), ggplot2::aes(x)) +
        ggplot2::geom_histogram(color = "black", fill = "white", bins = input$bins_stat) +
        ggplot2::geom_vline(xintercept = c(boot_res$ci.stat), color = "red", linetype = "dashed") +
        ggplot2::geom_vline(xintercept = boot_res$orig.stat, color = "blue") +
        ggplot2::labs(x = "Bootstrap Distribution", y = "Frequency")
    })

    # Effect Size Histogram
    output$hist_effect <- shiny::renderPlot({
      shiny::req(game_state()$analysis$bootstrap_results)

      # Bootstrap results
      boot_res <- game_state()$analysis$bootstrap_results

      ggplot2::ggplot(data.frame(x = boot_res$bootstrap.effect.dist), ggplot2::aes(x)) +
        ggplot2::geom_histogram(color = "black", fill = "white", bins = input$bins_effect) +
        ggplot2::geom_vline(xintercept = c(boot_res$ci.effect.size), color = "red", linetype = "dashed") +
        ggplot2::geom_vline(xintercept = boot_res$effect.size, color = "blue") +
        ggplot2::labs(x = "Bootstrap Distribution", y = "Frequency")
    })

    # 2. Conditional UI Rendering
    output$resultsUI <- shiny::renderUI({
      shiny::req(game_state()$current_round == input$rounds)

      list(
        shiny::h3("Results"),
        shiny::verbatimTextOutput("winner"),
        shiny::h3("Test Statistic"),
        shiny::tableOutput("test_stats"),
        shiny::plotOutput("hist_stat"),
        shiny::sliderInput("bins_stat", "Number of bins for Test Statistic:", min = 10, max = 50, value = 30),
        shiny::h3("Effect Size"),
        shiny::tableOutput("effect_size"),
        shiny::plotOutput("hist_effect"),
        shiny::sliderInput("bins_effect", "Number of bins for Effect Size:", min = 10, max = 50, value = 30)
      )
    })

  }

  shiny::shinyApp(ui = ui, server = server)

}
